content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
++++
+title = "Loop-Level Automatic Vectorization"
+[[extra.authors]]
+name = "Katy Voor"
+link = "https://github.com/kavoor"  
+[[extra.authors]]
+name = "Henry Liu"
+link = "https://www.linkedin.com/in/liuhenry4428/"  
++++
+
+## Introduction
+Modern processors have support for [SIMD](https://en.wikipedia.org/wiki/SIMD) instructions, which allow for efficient vector operations. We can leverage this feature to optimize loops that operate iteratively on arrays by changing operations that act on single array elements into vector operations that act on multiple array values in one instruction. 
+
+Consider the following loop of a vector-vector add:
+```gherkin=
+# Assume a, b, and c represent array base addresses in memory such
+# that the arrays do not overlap.
+...
+one: int = const 1;
+
+vvadd_loop:
+  ai: int = add a i;
+  bi: int = add b i;
+  ci: int = add c i;
+
+  va: int = lw ai;
+  vb: int = lw bi;
+  vc: int = add va vb;
+  sw vc ci;
+
+  i: int = add i one;
+  done: bool = ge i size;
+  br done vvadd_done vvadd_loop;
+```
+Compare this with the following loop:
+```gherkin=
+four: int = const 4;
+
+vvadd_loop:
+  ai: int = add a i;
+  bi: int = add b i;
+  ci: int = add c i;
+
+  va: int = vload ai;
+  vb: int = vload bi;
+  vc: int = vadd va vb;
+  vstore vc ci;
+
+  i: int = add i four;
+  done: bool = ge i size;
+  br done vvadd_done vvadd_loop;
+```
+The second loop executes for a forth the number of iterations of the first loop while behaving identically by using `vload`, `vstore`, and `vadd` that operate on four array elements at a time. This allows for *i* to be incremented by four each iteration instead of by one. 
```

Comment:
```suggestion
The second loop executes for a fourth the number of iterations of the first loop while behaving identically by using `vload`, `vstore`, and `vadd` that operate on four array elements at a time. This allows for *i* to be incremented by four each iteration instead of by one. 
```